### DISCUSSION TITLE
May I fork this codebase?

### DISCUSSION THREAD

**Discussion by mark-wiemer** [zero-plusplus/vscode-autohotkey-debug#332] [Q&A] [answered]
Hi @zero-plusplus, we've spoken earlier about incorporating your extension into an extension pack, and I respect your wish to keep your extension out of it. However, I do believe it offers great features and benefits to AHK coders, so I'd like to fork your codebase and make changes myself. This way, you wouldn't be seen as responsible for the changes, but of course I'd credit you as the original author. Does that seem reasonable?

Thanks,
Mark

**Comment by zero-plusplus** [comment, marked as answer]
Please stop with the fork.

What happened to what we talked about [here](https://github.com/mark-wiemer-org/autohotkey-extension-pack/issues/1) and [here](https://github.com/mark-wiemer-org/ahkpp/discussions/21#discussioncomment-9717144)?

And [here](https://github.com/mark-wiemer-org/ahkpp/discussions/21#discussioncomment-10228970) is what you said.

> I understand your position. I won't use your code in any way, don't worry :)

---

Do I need to explain it all again from scratch?

I am not developing for AHK coders. I develop for the users who install my extensions.

I put a lot of effort and time into getting users to choose my extensions.

But what if forked my program is incorporated into a larger extension?

What happens to the value of my extension?

I have explained it many times. But you, working for AHK coders, do not fully understand my thinking.

I again suggest that we keep our distance from each other as this difference in awareness creates problems.

---

PS

I feel you are looking for loopholes in my statement, so I will speak more clearly. (Yes, I did talk about copying and pasting, but not about forking.)

Please stop all actions that appropriates my work (I'll try to be vague here). This includes, for example, someone else forking my extension and you forking yours.

This is my intention, unlike a licence. It means that your actions are an impediment to my activities and I cannot accept them.

I will not allow anything in the future and will refuse to allow anything to be used without my permission.

I dare to speak out strongly because this is about work product.
But, I do not reject or condemn all of your activities.

It is simply that our ideas are not compatible.

> Doesn't make sense to duplicate our efforts.

Earlier, when asked about the state of parser development, you [said the above](https://github.com/zero-plusplus/vscode-autohotkey-debug/discussions/172#discussion-3708509).

However, the same effort can result in very different end products from one developer to another. 

So it is by no means in any sense meaningless, in my opinion. (Admitting this in the first place would mean that my activities are meaningless, as there are other people making AutoHotkey parsers than me.)

It seems like a waste of time to go down a path that someone else has already taken, but it's actually a lot of fun! :smile:

Thank you.

**Comment by jkyeung** [comment]
@zero-plusplus - I think Mark is acting in good faith and is _not_ looking for any loopholes. In fact, I think he has tried very hard to respect your wishes to the best of his ability, which includes trying to balance several considerations that sometimes overlap and sometimes conflict:

* Provide himself learning and enjoyment
* Provide value to the community (and I would argue this is more than just the "AHK community")
* Honor other people's explicit wishes

In my opinion, your position is difficult to understand, and therefore difficult to fully comply with. Your repo is public, and it is published under the MIT license. Those are deliberate choices that you made, and they are _fundamentally incompatible_ with your stated wishes and intentions.

I have a vague recollection that you may have tried to explain why you made those choices; and if you did, I think I might have eventually worked out some delicate, fragile understanding how you could reconcile those choices with your stated wishes and intentions. But it is hard to maintain that understanding over time. If you indeed had some written explanation, then you should _at minimum_ put that into the README of your repo, because without constant re-reading, and re-understanding, your position is simply illogical.

I strongly urge you to choose a different, more restrictive license that is aligned with your wishes. Keep in mind you can write your own license. You could look for an existing one that is close to what you want, and then modify it. In fact, I suspect that Mark would be willing to help you with that, since I truly believe he is trying to respect you and your work.

If all this sounds too difficult, or too much trouble, then it is really upon you to make your repo private instead of public. That might mean some of what you wanted to accomplish by having it public is no longer available to you. But the amount of time, effort, and emotional energy you are putting into defending your work suggests that the cost of having your work public is more than you can accept.

**Reply by zero-plusplus** [reply]
First, my argument is simple.
**”Please do not use my work to  (consequently) sabotage my activities.”**

---

My extension has a much smaller install count and community than AutoHotkey Plus Plus.

If my extension is fully included in AutoHotkey Plus Plus in such a situation, my extension will be worthless.

For example, have you ever seen someone in a game want to use a weapon that looks the same and is completely backwards compatible?

The above situation is like using my work to sabotage my activities. (At least that is how I perceive it.)

Are you saying that you think this is logical, rule-abiding, appropriate and respectful behavior for the author?
Does the MIT license require me to accept this?

When I chose the MIT License, I never imagined it would be used in this way.

> your position is simply illogical.

It is no wonder it is illogical.
I am about to lose the value of an extension that I have spent more than a few hundred hours growing in just a few clicks.

If I were logical I would not develop extensions with so much enthusiasm.
I worked hard because I had people supporting me.
It was just fun.
It's an extension that was born by emotion.

Before you wave the License around, take a look at the person in front of you.
Do I look happy?
Do you have respect for my activities?

I'm sorry if I didn't understand your intentions.
In that case, it seems I didn't understand you as you didn't understand me.

---

> I think Mark is acting in good faith

I agree with you here.

However, not all actions taken in good faith produce the right results.
What I am trying to say is that we are not compatible with me.

So I'm not suggesting that it's good or bad, but rather that we keep our distance without interfering with each other.

**Reply by jkyeung** [reply]
I don't think you understand what a license is.

The purpose of the license is to protect the author and the author's wishes. The license is where the author specifies what is acceptable and what is not acceptable use of their work. It's a contract, written by the author of the software. Users of the software are bound by the terms of that contract.

People who abide by the contract are not "waving it around" or using it as a weapon against you. They are using the software _according to what YOU have specified as acceptable_.

I really do not understand your analogy at all, but I will present one of my own:

Let's say you (the author) want to play a game of tennis. You post a sign that says "If you want to play tennis with me, then you must abide by the exact same rules that the Wimbledon tournament uses". Someone comes up to you and says, sure, I want to play tennis with you according to those rules. So you start to play. You serve the ball to the other player (the user). You like serving the ball, you like hitting the ball over the net, hitting the ball like that is fun for you. Great. Then the other player hits the ball back. And you suddenly say, "whoa, whoa, hold on there a minute! I don't want you to hit the ball back to me. What's fun is me hitting the ball to your side, and I really, really don't like you hitting the ball to my side. Please don't do that. OK, got it? If you respect me, then you will not hit the ball back. Now, let's resume the game according to the rules posted on the sign."

YOU chose the Wimbledon rules. They were just following YOUR rules.

If you don't like those rules, choose different ones! Write your own rules!

Your project really needs a different license. I would suggest you write down what you think is acceptable and what you think is not acceptable, and put that in the license. I suspect that what you write down will not be very common for GitHub projects, so I also _strongly recommend_ that you specifically point out in your README that this project uses a license of your own design, and so it is especially important for people to read the license carefully.

**Reply by zero-plusplus** [reply]
I see your point.
I have calmed down a bit. I know you're right because making exceptions to the license would create confusion.

Also, I understand that you are acting on behalf of Mark by my offensive words.
Let me apologize for that part.

I will take care not to be rude in the future, but there is a possibility that I may use inappropriate words due to the use of the translator. If this happens, please point it out to me.

My selfishness is to protect what is important to me in my own non-negotiable way, so I hope you can understand my thinking as much as possible.

Sorry about the license.
It was not intended to be used in such a way that it could be incorporated into other extensions because it could be linked to other extensions with a single install button.

So, the private project created on Mar 20, 2023, after understanding such concerns, states the following in the README.

> I will be developing this project with the goal of creating a standard development environment for AutoHotkey in the future. This repository contains a set of tools and editor extensions to make them happen.
>
> Therefore, please refrain from registering the this code itself or projects using it in the vscode marketplace, npm, or other registries.

So in conclusion, it is my fault for not assuming such a thing and not clearly stating my thoughts.

Fortunately we are in dialogue and would like to plug this hole if possible.

**Reply by mark-wiemer** [reply]
Thanks for the dialogue. For me, I don't understand why you see us as adversaries or enemies or competitors. We both want the same thing: a standard development environment for AHK. And I'm not sure why you say "I am not developing for AHK coders. I develop for the users who install my extensions." -- your extension is an AHK debugger, it's only going to provide value to folks that write AHK scripts, also known as AHK coders. Maybe you thought I meant folks that are updating the AHK language itself? Anyway, our audiences are the same: people that write AHK scripts.

I'm not looking for loopholes. I want to remind you that someone who is looking for loopholes has every legal right to fork your repo and republish it under their own name without even asking you. OK, that sounds weird written out--it's not a threat! I just want to clarify that I'm really going out of my way to make sure this is OK with you.

Four others have already forked your repo, which is why I thought you might be open to it.

Again, to be 100% clear, I will not fork your repo without your explicit permission, and I understand that I do not have that permission at this point.

But I think we can work together. We can add a funding page to AHK++ where **50% of the money goes to you (and 50% to thqby for his contributions)**--I'm a full-time engineer at Microsoft and don't need any extra cash right now. We can work out some sort of binding contract however you see fit. I want to support both you and the rest of the community, and I think working together with one big extension is better than two people working separately.

Your repo is very unique in the AHK community and is clearly the best AHK debugger I can find. I want to build off a solid foundation and I don't want to reinvent the wheel. I've done it many times before, and yes, it's fun, but it's also a lot slower. 😆

Again, I know that working together on a codebase can be tough as both of us would have to approve of code changes. Honestly, I'd probably create a submodule for your repo and you could make whatever changes you saw fit. I'd just add tests to make sure things didn't break unexpectedly. And I'd open pull requests for minor improvements as I encountered them. You'd keep full control over what happened within the submodule, I don't want to make coding any tougher.

Please consider it. I don't want to harm you or your work or the community in any way, I think us working together would be good for you, me, and everyone else. If you say no, that's totally OK, and I'll just build my own debugger with whoever wants to help. But I want to make a standard development environment for AutoHotkey--not 2 standard development environments.

**Reply by jkyeung** [reply]
@zero-plusplus - Don't worry, I do not think you were rude, and I was not offended by anything you said. I also don't believe you did or said anything that you need to apologize for. I'm glad for the dialogue.

**Reply by mark-wiemer** [reply]
I agree with @jkyeung on just about everything. You don't need to apologize--you're stating your thoughts. And I second their advice for a new license. The MIT license is **very** permissive, and other folks will reasonably see it and think they can fork your repo without asking. I would strongly suggest simply making this repo private so that others do not copy it, and then work to find or write a license that suits your needs before making it public again.

**Reply by zero-plusplus** [reply]
> I don't want to reinvent the wheel

You described my extension as unique, but most of it is just reinventing the wheel. It was tedious and painful for me, but necessary.

> I am not developing for AHK coders. I develop for the users who install my extensions.

This was to express the difference in the size of the communities we were looking at.

You are referring to the entire AHK user base, I am referring to the AHK users who utilize my extension.

> I want to create a standard development environment for AutoHotkey, not two standard development environments.

I want to create my own development environment.
This is one of my dreams and a problem I would like to solve with my own hands.

---

Unfortunately we can't seem to understand each other.

I enjoy competition in extension development.
I understand your stance, but if you apply it directly to me, I lose one of the pleasures.

It will certainly take several years to surpass your extension.

That's why I enjoyed the position I'm in now, where I've been able to grow freely while looking several years into the future. I never thought that this position could be easily shaken by a single forked proposal.

You are correct, there is no problem in terms of licensing. My words mean nothing in the MIT license.

So I am going to throw out the MIT license and start over from scratch with a new license that does not waive all copyrights in the new repository.

**Reply by zero-plusplus** [reply]
> Don't worry, I do not think you were rude, and I was not offended by anything you said. I also don't believe you did or said anything that you need to apologize for. I'm glad for the dialogue.

Thank you. I'm glad we could have a dialogue too.

**Reply by mark-wiemer** [reply]
OK, I see, you want to create your own environment, not one that everyone can collaborate on and improve. I understand that--for you this is a personal project, and that's totally OK. I'm glad you're starting anew without the MIT license, that will certainly resolve questions like this in the future.

Well, if anyone would like to join me, I'll be writing my own advanced debugger for AHK now over at [AHK++](https://github.com/mark-wiemer-org/ahkpp) :)

**Reply by zero-plusplus** [reply]
A few days have passed and I have cooled down again.

I apologize again for my lack of understanding and assumptions about licensing that had created a major misunderstanding.

My objectives and the license's objectives were very different. It was only natural that our conversations should have crossed paths, as my fault had caused our assumptions about each other to be so far off.

I will be careful not to repeat the same mistake by making my objectives and licenses as consistent and clearer as possible in new project.

I really hope to see your extension grow and develop as your debugger and integrated development environment.

I will try to develop an independent extension that will be attractive and noteworthy in the new project for my own enjoyment.

After posting this comment, I will double-check and correct for any disrespectful content in the maintenance stop reports, etc. that I may have written before my head cooled down. I am truly sorry.

Finally, I will be temporarily disabling my Github sponsorship going forward. As long as the sponsor is still in place, it seems that it cannot be disabled, so I apologize for the inconvenience, please discontinue the continuation of the sponsorship.

---

I have reviewed it many times and I hope you didn't find anything wrong with my writing.
